### ISSUE TITLE
Problem with "super-marshaller" for various content types

### ISSUE THREAD

**Issue by Susak** [akka/akka-http#1017]
For example I have two marshallers for different content-types, which provide marshalling for MyClass to String.
I have run the following code. 
```
import akka.http.scaladsl.marshalling.Marshaller
import akka.http.scaladsl.model.{HttpRequest, MediaTypes}
import akka.http.scaladsl.model.headers.Accept

object MarshallerTest {
  def main(args: Array[String]): Unit = {
    import scala.concurrent.ExecutionContext.Implicits.global

    case class MyClass(value: String)

    implicit val superMarshaller = {
      val jsonMarshaller =
        Marshaller.stringMarshaller(MediaTypes.`application/json`)
          .compose[MyClass] { mc =>
          println(s"jsonMarshaller marshall $mc")
          mc.value
        }
      val textMarshaller = Marshaller.stringMarshaller(MediaTypes.`text/html`)
        .compose[MyClass] { mc =>
        println(s"textMarshaller marshall $mc")
        throw new IllegalArgumentException(s"Unexpected value $mc")
      }

      Marshaller.oneOf(jsonMarshaller, textMarshaller)
    }
    val request =
      HttpRequest(uri = "/test")
        .withHeaders(Accept(MediaTypes.`application/json`))
    akka.http.scaladsl.marshalling.Marshal(MyClass("test"))
      .toResponseFor(request) onComplete println
  }
}
```
Marshaller `jsonMarshaller` marshall for `application/json` MediaType and marshaller `textMarshaller` do same for  `text/html` MediaType.
I marshall `MyClass("test")` to response for `request` with `Accept: application/json` header, in my opinion here should choose `jsonMarshaller` and marshalling should be successful.
But before you find the right implementation we call apply method here https://github.com/akka/akka-http/blob/master/akka-http/src/main/scala/akka/http/scaladsl/marshalling/Marshal.scala#L42
and then apply value for both implementations https://github.com/akka/akka-http/blob/master/akka-http/src/main/scala/akka/http/scaladsl/marshalling/Marshaller.scala#L92
Can you please explain is it feature or bug?

**Comment by ktoso** [comment]
It is how the current infra works, yes - it invokes both indeed.

There is a ticket to change it, if we can: https://github.com/akka/akka-http/issues/243

I've added some more docs about this in: https://github.com/akka/akka-http/pull/1018

**Comment by ktoso** [comment]
I'll close this as docs issue, but generally speaking it's duplicate of 243.

**Comment by Susak** [comment]
Thanks for the quick response :)
I've  led is the contrived example.
But when we have few marshallers with  complex calculations, whose we pass to Marshaller.oneOf,  it will be bad for performance, if invoke all of them.

**Comment by ktoso** [comment]
That's why that other ticket exists, no one is disagreeing with that.
Look at the types, it's impossible to not allocate the Marshalling right
now.

This ticket will be closed once the scaladoc is merged, the improvement
under discussion is the other ticket I have linked.

-- 
Konrad `ktoso` Malawski
Akka <http://akka.io> @ Lightbend <http://lightbend.com>

On 6 April 2017 at 21:36:20, Ruslan Sokolov (notifications@github.com)
wrote:

Thanks for the quick response :)
I've led is the contrived example.
But when we have few marshallers with complex calculations, whose we pass
to Marshaller.oneOf, it will be bad for performance, if invoke all of them.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/akka/akka-http/issues/1017#issuecomment-292289850>, or mute
the thread

**Comment by ktoso** [comment]
For the marshalling infra change please track https://github.com/akka/akka-http/issues/243
